stellargraph/mapper/knowledge_graph.py

```diff
@@ -0,0 +1,189 @@
+
+
+class KGTripleGenerator:
+    """
+    A data generator for working with triple-based knowledge graph models, like ComplEx.
+
+    This requires a StellarGraph that contains all nodes/entities and every edge/relation type that
+    will be trained or predicted upon. The graph does not need to contain the edges/triples that are
+    used for training or prediction.
+
+    Args:
+        G (StellarGraph): the graph containing all nodes, and all edge types.
+
+        batch_size (int): the size of the batches to generate
+    """
+
+    def __init__(self, G, batch_size):
+        self.G = G
+
+        if not isinstance(batch_size, int):
+            raise TypeError(
+                f"batch_size: expected int, found {type(batch_size).__name__}"
+            )
+
+        self.batch_size = batch_size
+
+    def flow(self, edges, negative_samples=None, shuffle=False, seed=None):
+        """
+        Create a Keras Sequence yielding the edges/triples in ``edges``, potentially with some negative
+        edges.
+
+        The negative edges are sampled using the "local closed world assumption", where a
+        source/subject or a target/object is randomly mutated.
+
+        Args:
+            edges: the edges/triples to feed into a knowledge graph model.
+            negative_samples (int, optional): the number of negative samples to generate for each positive edge.
+
+        Returns:
+            A Keras sequence that can be passed to the ``fit`` and ``predict`` method of knowledge-graph models.
+        """
+        if isinstance(edges, pd.DataFrame):
+            sources = edges[SOURCE]
+            rels = edges[TYPE_ATTR_NAME]
+            targets = edges[TARGET]
+        else:
+            raise TypeError(
+                f"edges: expected pandas.DataFrame; found {type(edges).__name__}"
+            )
+
+        if negative_samples is not None:
+            if not isinstance(negative_samples, int):
+                raise TypeError(
+                    f"negative_samples: expected int or None, found {type(negative_samples).__name__}"
+                )
+            if negative_samples < 0:
+                raise ValueError(
+                    f"negative_samples: expected non-negative integer, found {negative_samples}"
+                )
+
+        source_ilocs = self.G._get_index_for_nodes(sources)
+        # FIXME(#870): this would be best expressed without poking into the _edges proprety of G
+        rel_ilocs = self.G._edges.types.to_iloc(rels, strict=True)
+        target_ilocs = self.G._get_index_for_nodes(targets)
+
+        return KGTripleSequence(
+            max_node_iloc=self.G.number_of_nodes(),
+            source_ilocs=source_ilocs,
+            rel_ilocs=rel_ilocs,
+            target_ilocs=target_ilocs,
+            batch_size=self.batch_size,
+            shuffle=shuffle,
+            negative_samples=negative_samples,
+            seed=seed,
+        )
+
+
+class KGTripleSequence(Sequence):
```

Comment:
Could this class be replaced with a tensorflow `Dataset`?

Something along the lines of:
`dataset = tf.data.Dataset.from_tensor_slices([source_ilocs, rel_ilocs, target_ilocs]).shuffle().batch(batch_size).map(negative_sampling_batch_function)` 

Would do the trick. This should be faster, allow multithreading, and would remove the boiler plate batch shuffling code.

Reply:
sorry typo above, "multithreading" should've been "c multiprocessing" - which I believe has less overhead than python multiprocessing.

Reply:
(As above, https://github.com/stellargraph/stellargraph/issues/876)